src/core/thread/mle_router.cpp

```diff
@@ -4476,6 +4477,66 @@ Error MleRouter::SendTimeSync(void)
 }
 #endif // OPENTHREAD_CONFIG_TIME_SYNC_ENABLE
 
+Error MleRouter::DetachGracefully(otDetachGracefullyCallback aCallback, void *aContext)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(aCallback != nullptr, error = kErrorInvalidArgs);
+
+    if (IsDisabled() || IsDetached())
+    {
+        ExitNow(error = kErrorAlready);
+    }
+
+    if (IsChild())
+    {
+        ExitNow(error = Mle::DetachGracefully(aCallback, aContext));
+    }
+
+    if (IsLeader())
+    {
+        ExitNow(error = kErrorNone);
+    }
+
+    OT_ASSERT(IsRouter());
+
+    mDetachGracefullyCallback = aCallback;
+    mDetachGracefullyContext  = aContext;
+
+    SendAddressRelease(&MleRouter::HandleDetachGracefullyAddressReleaseResponse, this);
```

Comment:
Can we still handle address release in the original handler? 
For example, we can check if the `mDetachGracefullyTimer` is set to determine if the device is detaching gracefully.

Reply:
There is no original handler here. `SendAddressRelease` doesn't set a handler before this change.